### PR TITLE
BiRNN refactored, tested and doc updated, other minor updates to rnn, fixed dropout bug, make format

### PR DESCRIPTION
**Pull request by JingqingZ**
### Checklist
- [x] I've tested that my changes are compatible with the latest version of Tensorflow.
- [x] I've read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### CODE REVIEW

**Comment by ChrisWu1997** [comment]
test_lambda part reviewed. For RNN part, @zsdonghao could you please review it? I'm not very familiar with RNN.

**Comment by zsdonghao** [comment]
@JingqingZ so, we would not support `n_layer` and `dropout` in RNNs now?

**Comment by JingqingZ** [comment]
> @JingqingZ so, we would not support `n_layer` and `dropout` in RNNs now?

`dropout` is supported in `cells`.

Users can stack `RNN` and `BiRNN` to achieve the functionality of `n_layer` in TL1.X.

**Comment by JingqingZ** [comment]
Bug in `dropout` in rnn cell.

`dropout` may not be activated in training or deactivated in testing. Tests needed.

**Comment by JingqingZ** [comment]
Please check add a new test in [tests/layers/test_layers_core_basedense_dropout.py](https://github.com/zsdonghao/tensorlayer2/blob/master/tests/layers/test_layers_core_basedense_dropout.py) scenario:
```python
def test_dropout(self):
        data_x = np.random.random([10, 784]).astype(np.float32)
        pred_y_1 = self.model(data_x, is_train=True)
        pred_y_2 = self.model(data_x, is_train=True)
        print(pred_y_1)
        print(pred_y_2)
        self.assertFalse(np.allclose(pred_y_1, pred_y_2))
        pred_y_1 = self.model(data_x, is_train=False)
        pred_y_2 = self.model(data_x, is_train=False)
        print(pred_y_1)
        print(pred_y_2)
        self.assertTrue(np.allclose(pred_y_1, pred_y_2))
```

It seems `tf.function` will ignore the change of `self.is_train`. For example, the forward function of `Dropout` layer depends on `self.is_train`, but no matter `model.eval()` or `model.train()`, the `self.is_train` in `Dropout` is always `True` (because in `Layer` core, `self.is_train` is initialized as `True`).

**Comment by ChrisWu1997** [comment]
We should remove all @tf.function in layer forward function and let user add @tf.function for their training function as tensorflow does. Although I don't like this design, it seems to be all we can do now.

**Comment by JingqingZ** [comment]
The `dropout` bug in rnn/birnn cell has been solved. Please have a check and review. @ChrisWu1997 @zsdonghao
